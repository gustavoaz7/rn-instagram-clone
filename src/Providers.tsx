import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider } from 'react-redux';
import { theme as stylesTheme } from './styles/theme';
import { store as reduxStore } from './redux/store';

type TProvidersProps = {
  children: ReactNode;
  store?: typeof reduxStore;
  theme?: typeof stylesTheme;
};

export function Providers({
  children,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>
  );
}
