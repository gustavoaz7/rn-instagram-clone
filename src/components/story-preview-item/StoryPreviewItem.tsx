import React, { useEffect, useRef, useState } from 'react';
import { Animated } from 'react-native';
import { LinearGradient } from 'expo-linear-gradient';
import styled from 'styled-components/native';
import { Text } from '../text';

export const SCALE_DURATION = 150;

export function StoryPreviewItem(): JSX.Element {
  const [isPressing, setIsPressing] = useState(false);
  const scale = useRef(new Animated.Value(1)).current;

  useEffect(() => {
    Animated.timing(scale, {
      toValue: isPressing ? 0.9 : 1,
      duration: SCALE_DURATION,
      useNativeDriver: true,
    }).start();
  }, [isPressing, scale]);

  return (
    <Container
      onPressIn={() => setIsPressing(true)}
      onPressOut={() => setIsPressing(false)}
    >
      <RingContainer style={{ transform: [{ scale }] }}>
        <StatusRing>
          <AvatarContainer>
            <Avatar
              source={{
                uri: 'https://cdn.fakercloud.com/avatars/spbroma_128.jpg',
              }}
            />
          </AvatarContainer>
        </StatusRing>
      </RingContainer>
      <Text numberOfLines={1}>veryLongNameHere</Text>
    </Container>
  );
}

const Container = styled.Pressable`
  align-items: center;
  width: 76px;
`;

const RingContainer = styled(Animated.View)`
  position: relative;
  height: 74px;
  width: 74px;
  border-radius: 74px;
  overflow: hidden;
`;

const StatusRing = styled(LinearGradient).attrs(({ theme }) => ({
  colors: [theme.color.purpleRed, theme.color.yellow],
  start: { x: 0.7, y: 0.3 },
  end: { x: 0.3, y: 0.7 },
}))`
  width: 100%;
  height: 100%;
`;

const AvatarContainer = styled.View`
  position: absolute;
  top: 2px;
  left: 2px;
  width: 70px;
  height: 70px;
  border-radius: 70px;
  background-color: ${({ theme }) => theme.color.white};
`;

const Avatar = styled.Image`
  position: absolute;
  top: 3px;
  left: 3px;
  width: 64px;
  height: 64px;
  border-radius: 64px;
`;
