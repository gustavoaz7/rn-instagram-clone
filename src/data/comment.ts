import { TComment } from '../types';

export const generateComment = (): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  likedBy: ['Frankie30', 'Patricia24', 'Westley_Ebert49'],
});
