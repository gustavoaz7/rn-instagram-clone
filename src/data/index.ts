import type {
  TLike,
  TComment,
  TPostMedia,
  TPost,
  TOwner,
  TStory,
  TStoryMedia,
  TProfile,
  TTappableObject,
} from '../types';

export const generateMockOwner = (): TOwner => ({
  id: `${Math.random()}`,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
});

export const generateMockLike = (): TLike => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
});

export const generateMockComment = (likesQty = 1): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  viewerHasLiked: false,
});

export const generateMockTappableObject = ({
  type = 'mention',
  text = 'random',
  x = 0.1,
  y = 0.2,
  scale = 1.1,
  rotation = 300,
}: Partial<TTappableObject> = {}): TTappableObject => ({
  id: `${Math.random()}`,
  type,
  text,
  x,
  y,
  scale,
  rotation,
});

export const generateMockPostMedia = ({ index = 0 } = {}): TPostMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
});

export const generateMockPost = ({
  mediasQty = 1,
  likesQty = 3,
  commentsQty = 1,
} = {}): TPost => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
  medias: [...Array(mediasQty)].map(generateMockPostMedia),
  caption:
    'Tempora rerum minima dolor quibusdam blanditiis ea et. Earum iste iure expedita velit laudantium inventore quia qui. Distinctio consectetur accusamus voluptatum hic rem at quas. Incidunt ipsum sed illo molestiae vitae ullam sit consequatur autem. Ut voluptatem nisi. Velit ex dignissimos sed aut.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  previewComments: {
    count: commentsQty,
    comments: [...Array(commentsQty)].map(generateMockComment),
  },
  location: 'Daronville',
  viewerHasLiked: false,
});

export const generateMockStoryMedia = ({
  viewersQty = 0,
  index = 0,
} = {}): TStoryMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
  previewViewers: {
    count: viewersQty,
    viewers: [...Array(viewersQty)].map(generateMockOwner),
  },
  takenAt: 1620488304326,
  expiresAt: 1620488304326 + 3600 * 24,
});

export const generateMockStory = ({
  mediasQty = 3,
  viewersQty = 0,
} = {}): TStory => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  expiresAt: 0,
  latestMediaAt: 0,
  seenAt: 0,
  medias: [...Array(mediasQty)].map((_, index) =>
    generateMockStoryMedia({ viewersQty, index }),
  ),
});

export const generateMockProfile = (): TProfile => ({
  bio: 'animi eum earum aut explicabo',
  followCount: 47,
  followedByCount: 50,
  followedByViewer: false,
  followsViewer: false,
  fullName: 'Gustavo Az',
  postsCount: 20,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
  story: generateMockStory(),
});
