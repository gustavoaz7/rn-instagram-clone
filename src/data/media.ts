import type { TPostMedia } from '../types';

export const generatePostMedia = (): TPostMedia => ({
  id: `${Math.random()}`,
  url: 'http://placeimg.com/640/480/cats',
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
    username: 'gustavoaz7',
  },
  tappableObjects: [],
});
