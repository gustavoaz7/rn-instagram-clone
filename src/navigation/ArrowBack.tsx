import React from 'react';
import { useNavigation } from '@react-navigation/native';
import styled, { useTheme } from 'styled-components/native';
import ArrowLeft from '../../assets/svg/arrow-left.svg';

export function ArrowBack(): JSX.Element {
  const navigation = useNavigation();
  const theme = useTheme();

  return (
    <ArrowContainer onPress={navigation.goBack}>
      <ArrowLeft color={theme.color.foreground} />
    </ArrowContainer>
  );
}

const ArrowContainer = styled.Pressable`
  margin-left: ${({ theme }) => theme.spacing.l};
`;
