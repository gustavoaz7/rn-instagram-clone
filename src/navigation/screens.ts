export enum BOTTOM_TAB_SCREENS {
  HOME = '@Bottom-Tab/Home',
  SEARCH = '@Bottom-Tab/Search',
  REELS = '@Bottom-Tab/Reels',
  SHOP = '@Bottom-Tab/Shop',
  PROFILE = '@Bottom-Tab/Profile',
}

export enum DIRECT_STACK_SCREENS {
  DIRECT = '@Direct-Stack/Direct',
  VIDEO = '@Direct-Stack/Video',
  MESSAGE = '@Direct-Stack/Message',
  CHAT = '@Direct-Stack/Chat',
}

export enum HOME_STACK_SCREENS {
  FEED = '@Home-Stack/Feed',
  ACTIVITY = '@Home-Stack/Activity',
  PROFILE = '@Home-Stack/Profile',
}

export enum HOME_SWIPE_SCREENS {
  CAMERA = '@Home-Swipe/Camera',
  BOTTOM_TAB = '@Home-Swipe/Bottom-Tab',
  DIRECT = '@Home-Swipe/Direct',
}

export enum ROOT_STACK_SCREENS {
  HOME_SWIPE = '@Root-Stack/Home-Swipe',
  COMMENTS = '@Root-Stack/Comments',
  STORY = '@Root-Stack/Story',
}
