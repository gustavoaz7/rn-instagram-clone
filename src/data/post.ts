import { TPost } from '../types';
import { generatePostMedia } from './media';

export const createMockPost = (): TPost => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
    username: 'gustavoaz7',
  },
  createdAt: 1620488304326,
  medias: [generatePostMedia()],
  caption:
    'Tempora rerum minima dolor quibusdam blanditiis ea et. Earum iste iure expedita velit laudantium inventore quia qui. Distinctio consectetur accusamus voluptatum hic rem at quas. Incidunt ipsum sed illo molestiae vitae ullam sit consequatur autem. Ut voluptatem nisi. Velit ex dignissimos sed aut.',
  likedBy: ['Tito.Zieme', 'Cleo.Runte69', 'Laverna.Emmerich24'],
  comments: [
    {
      id: `${Math.random()}`,
      owner: {
        id: `${Math.random()}`,
        profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
        username: 'Vivienne_Doyle',
      },
      createdAt: 1620488304326,
      text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
      likedBy: ['Frankie30', 'Patricia24', 'Westley_Ebert49'],
    },
  ],
  location: 'Daronville',
});
