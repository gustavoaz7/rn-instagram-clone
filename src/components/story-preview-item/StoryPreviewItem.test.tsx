import React from 'react';
import {
  render,
  RenderOptions,
  fireEvent,
} from '@testing-library/react-native';
import { Animated } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { StoryPreviewItem, SCALE_DURATION } from './StoryPreviewItem';
import { Providers } from '../../Providers';
import {
  setupTimeTravel,
  destroyTimeTravel,
  timeTravel,
} from '../../test/time-travel';
import { generateMockOwner } from '../../data';
import { ROOT_STACK_SCREENS } from '../../navigation/screens';

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));
const useNavigationMock = useNavigation as jest.Mock;

describe('components - StoryPreviewItem', () => {
  const owner = generateMockOwner();
  const options: RenderOptions = { wrapper: Providers };
  const navigateSpy = jest.fn();
  useNavigationMock.mockReturnValue({ navigate: navigateSpy });

  beforeEach(() => {
    setupTimeTravel();
    navigateSpy.mockReset();
  });

  afterEach(() => {
    destroyTimeTravel();
  });

  it('renders', () => {
    render(<StoryPreviewItem owner={owner} />, options);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<StoryPreviewItem owner={owner} />, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it('bounces on press', () => {
    const { UNSAFE_getByType } = render(
      <StoryPreviewItem owner={owner} />,
      options,
    );

    const element = UNSAFE_getByType(Animated.View);
    const animatedScale = element.props.style.transform[0].scale;

    fireEvent(element, 'pressIn');
    timeTravel(SCALE_DURATION);

    expect(animatedScale).toMatchInlineSnapshot('0.9');

    fireEvent(element, 'pressOut');
    timeTravel(SCALE_DURATION);

    expect(animatedScale).toMatchInlineSnapshot('1');
  });

  it('navigates to story screen on press', () => {
    const { getByTestId } = render(<StoryPreviewItem owner={owner} />, options);

    fireEvent.press(getByTestId('StoryPreviewItem'));

    expect(navigateSpy).toHaveBeenCalledTimes(1);
    expect(navigateSpy).toHaveBeenCalledWith(ROOT_STACK_SCREENS.STORY, {
      username: owner.username,
    });
  });
});
