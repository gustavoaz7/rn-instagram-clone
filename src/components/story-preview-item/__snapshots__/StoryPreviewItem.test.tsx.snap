// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - StoryPreviewItem matches snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "width": 76,
      },
    ]
  }
  testID="StoryPreviewItem"
>
  <View
    style={
      Object {
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      }
    }
  >
    <View
      size={74}
      style={
        Array [
          Object {
            "borderRadius": 74,
            "height": 74,
            "overflow": "hidden",
            "position": "relative",
            "width": 74,
          },
        ]
      }
    >
      <ViewManagerAdapter_ExpoLinearGradient
        proxiedProperties={
          Object {
            "colors": Array [
              4292948076,
              4294750021,
            ],
            "endPoint": Array [
              0.3,
              0.7,
            ],
            "locations": undefined,
            "startPoint": Array [
              0.7,
              0.3,
            ],
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
      >
        <View
          ringWidth={2}
          size={74}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 70,
                "height": 70,
                "justifyContent": "center",
                "width": 70,
              },
            ]
          }
        >
          <Image
            offset={3}
            ringWidth={2}
            size={74}
            source={
              Object {
                "uri": "https://avatars.githubusercontent.com/u/28453820?v=4",
              }
            }
            style={
              Array [
                Object {
                  "borderColor": "#26262633",
                  "borderRadius": 64,
                  "borderWidth": 2,
                  "height": 64,
                  "width": 64,
                },
              ]
            }
          />
        </View>
      </ViewManagerAdapter_ExpoLinearGradient>
    </View>
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#262626",
        },
      ]
    }
  >
    gustavoaz7
  </Text>
</View>
`;
